girder/models/model_base.py

```diff
@@ -1511,10 +1581,52 @@ def prefixSearch(self, query, user=None, filters=None, limit=0, offset=0,
         :returns: A pymongo cursor. It is left to the caller to build the
             results from the cursor.
         """
-        filters = filters or {}
+        filters = self._prefixSearchFilters(query, filters, prefixSearchFields)
+
+        return self.findWithPermissions(
+            filters, offset=offset, limit=limit, sort=sort, fields=fields,
+            user=user, level=level)
+
+    def permissionClauses(self, user=None, level=None, prefix=''):
+        return _permissionClauses(user, level, prefix)
+
+    def findWithPermissions(self, query=None, offset=0, limit=0, timeout=None, fields=None,
+                            sort=None, user=None, level=AccessType.READ, **kwargs):
+        """
+        Search the collection by a set of parameters, only returning results
+        that the combined user and level have permission to access. Passes any
+        extra kwargs through to the underlying pymongo.collection.find
+        function.
 
-        cursor = Model.prefixSearch(
-            self, query, filters=filters, sort=sort, fields=fields,
-            prefixSearchFields=prefixSearchFields)
-        return self.filterResultsByPermission(
-            cursor, user=user, level=level, limit=limit, offset=offset)
+        :param query: The search query (see general MongoDB docs for "find()")
+        :type query: dict
+        :param offset: The offset into the results
+        :type offset: int
+        :param limit: Maximum number of documents to return
+        :type limit: int
+        :param timeout: Cursor timeout in ms. Default is no timeout.
+        :type timeout: int
+        :param fields: A mask for filtering result documents by key, or None to return the full
+            document, passed to MongoDB find() as the `projection` param.
+        :type fields: `str, list of strings or tuple of strings for fields to be included from the
+            document, or dict for an inclusion or exclusion projection`.
+        :param sort: The sort order.
+        :type sort: List of (key, order) tuples.
+        :param user: The user to check policies against.
+        :type user: dict or None
+        :param level: The access level.  Explicitly passing None skips doing
+            permissions checks.
+        :type level: AccessType
+        :returns: A pymongo Cursor or CommandCursor.  If a CommandCursor, it
+            has been augmented with a count function.
+        """
+        # If no user is specified and greater than READ access is requested,
+        # we won't return anything.  So as to always return a cursor for
+        # consistency, we make a query that will return no results
+        if user is None and level is not None and level > AccessType.READ:
+            query = {'__matchnothing': 'nothing'}
```

Comment:
Is there a better way to fake this that won't hit the database at all? I worry that a query like this might actually be slower than leaving this condition out, but I'd have to do some reading/experiments to be sure.

Reply:
The virtue of faking it this way is that we return an actual pymongo `Cursor`, so it behaves as expected.  I'm sure it takes more time than returning a iterator with a synthetic count function, but it is probably a fairly rare condition -- no user asking for write or better access.  Returning something else means that other code could need a special code path.

The other alternative would be to generate a Mongo Cursor that is more faked.  For instance, after the `self.find()` returns a Cursor, it could be monkey-patched by reaching in and saying `<cursor>._Cursor__empty = True` or perhaps calling `<cursor>._Cursor__die()`, which would prevent any round-trip to Mongo from occurring.  This seems hackier (and less stable -- though it could be done if the `_Cursor__die` function exists and not if it doesn't).